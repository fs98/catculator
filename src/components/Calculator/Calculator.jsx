import { useMemo, useState } from "react";
import { useEffect } from "react/cjs/react.development";
import Button from "../Button/Button";

const buttonsList = [
  { title: "7", type: "number" },
  { title: "8", type: "number" },
  { title: "9", type: "number" },
  { title: "/", type: "action" },
  { title: "4", type: "number" },
  { title: "5", type: "number" },
  { title: "6", type: "number" },
  { title: "x", type: "action" },
  { title: "1", type: "number" },
  { title: "2", type: "number" },
  { title: "3", type: "number" },
  { title: "-", type: "action" },
  { title: "AC", type: "action" },
  { title: "0", type: "number" },
  { title: "=", type: "action" },
  { title: "+", type: "action" },
];

const Calculator = () => {
  const [valueA, setValueA] = useState(null);
  const [valueB, setValueB] = useState(null);
  const [action, setAction] = useState(null);
  const [totalResult, setTotalResult] = useState(0);
  const [currentValue, setCurrentValue] = useState(0);

  useEffect(() => {
    console.log("valueA", valueA);
    console.log("valueB", valueB);
    console.log("action", action);
  }, [action, valueA, valueB]);

  useEffect(() => {
    if (valueA === null) {
      setCurrentValue(0);
    }

    if (valueB === null) {
      setCurrentValue(valueA);
    }

    if (action !== null) {
      setCurrentValue(totalResult);
    }

  }, [action, totalResult, valueA, valueB]);

  const resultHandler = () => {
    // if (valueA === null || valueB === null || action === null) return;

    if ([valueA, valueB, action].includes(null)) return;

    switch (action) {
      case "/":
        setTotalResult(valueA / valueB);
        break;
      case "x":
        setTotalResult(valueA * valueB);
        break;
      case "+":
        setTotalResult(valueA + valueB);
        break;
      case "-":
        setTotalResult(valueA - valueB);
        break;
      default:
        return 0;
    }
  };

  const onButtonClickHandler = (value) => {
    if (value.type === "number") {
      if (valueA === null || action === null) {
        setValueA(Number(value.title));
      } else {
        setValueB(Number(value.title));
      }
    } else {
      if (value.title === "=") {
        resultHandler();
      } else if (value.title === "AC") {
        setValueA(null);
        setValueB(null);
        setAction(null);
      } else {
        setAction(value.title);
      }
    }
  };

  return (
    <div className="calculator container mx-auto">
      <h1 className="py-8 px-8 text-right text-6xl">{currentValue}</h1>

      <div className="grid grid-cols-4 gap-2">
        {buttonsList.map((btn, index) => {
          return (
            <Button
              key={btn.title}
              type={btn.type}
              onClick={() => onButtonClickHandler(btn)}
            >
              {btn.title}
            </Button>
          );
        })}
      </div>
    </div>
  );
};

export default Calculator;
