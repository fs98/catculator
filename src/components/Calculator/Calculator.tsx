import { useState, useEffect } from "react";
import { Button } from "../Button/Button";

export type ActionProps = "/" | "x" | "+" | "-" | "AC" | "=";
export type ButtonType = "number" | "action";
export type ButtonListProps = {
  title: ActionProps | number;
  type: ButtonType;
};

const buttonsList: ButtonListProps[] = [
  { title: 7, type: "number" },
  { title: 8, type: "number" },
  { title: 9, type: "number" },
  { title: "/", type: "action" },
  { title: 4, type: "number" },
  { title: 5, type: "number" },
  { title: 6, type: "number" },
  { title: "x", type: "action" },
  { title: 1, type: "number" },
  { title: 2, type: "number" },
  { title: 3, type: "number" },
  { title: "-", type: "action" },
  { title: "AC", type: "action" },
  { title: 0, type: "number" },
  { title: "=", type: "action" },
  { title: "+", type: "action" },
];

export const Calculator = (): JSX.Element => {
  const [valueA, setValueA] = useState<number>();
  const [valueB, setValueB] = useState<number>();
  const [action, setAction] = useState<ActionProps>();
  const [totalResult, setTotalResult] = useState<number>();
  const [currentValue, setCurrentValue] = useState<number>(0);

  useEffect(() => {
    if (valueA === undefined) {
      setCurrentValue(0);
    } else if (valueB === undefined) {
      setCurrentValue(valueA);
    } else {
      setCurrentValue(valueB);
    }
  }, [valueA, valueB]);

  useEffect(() => {
    if (totalResult !== undefined) {
      setCurrentValue(totalResult);
    }
  }, [totalResult]);

  const resultHandlerSwitch = (result: number) => {
    setTotalResult(result);
    setValueA(result);
  };

  const resultHandler = () => {
    // if ([valueA, valueB, action].includes(undefined)) return;
    if (valueA === undefined || valueB === undefined || action === undefined)
      return;

    // We never reseted the A and B values so when we choose second action it always adds up to B string value, and A value stays the same. This way, after choosing action "=" the total result becomes A, and B is empty.

    switch (action) {
      case "/":
        resultHandlerSwitch(valueA / valueB);
        break;
      case "x":
        resultHandlerSwitch(valueA * valueB);
        break;
      case "+":
        resultHandlerSwitch(valueA + valueB);
        break;
      case "-":
        resultHandlerSwitch(valueA - valueB);
        break;
      case undefined:
      default:
        setTotalResult(0);
    }
    setValueB(undefined);
  };

  const onButtonClickHandler = (value: ButtonListProps) => {
    if (value.type === "number") {
      if (action === undefined) {
        setValueA((prev) =>
          Number(
            !!prev
              ? prev.toString() + value.title.toString()
              : value.title.toString()
          )
        );
      } else {
        setValueB((prev) =>
          Number(
            !!prev
              ? prev.toString() + value.title.toString()
              : value.title.toString()
          )
        );
      }
    } else {
      if (value.title === "=") {
        resultHandler();
      } else if (value.title === "AC") {
        setValueA(undefined);
        setValueB(undefined);
        setAction(undefined);
        setTotalResult(0);
      } else if (["+", "-", "/", "x"].includes(value.title.toString())) {
        if (currentValue === 0) {
          setValueA(0);
        }
        setAction(value.title as ActionProps);
      }
    }
  };

  return (
    <div className="calculator container bg-black bg-opacity-40 rounded-md p-4 mx-auto md:w-1/2">
      <h1 className="p-8 text-right text-white text-6xl">{currentValue}</h1>

      <div className="grid grid-cols-4 gap-2">
        {buttonsList.map((btn, index) => {
          return (
            <Button
              key={btn.title}
              type={btn.type}
              onClick={() => onButtonClickHandler(btn)}
            >
              {btn.title}
            </Button>
          );
        })}
      </div>
    </div>
  );
};

export default Calculator;
