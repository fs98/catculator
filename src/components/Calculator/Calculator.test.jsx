import {
  render,
  screen,
  fireEvent,
  waitFor,
} from "@testing-library/react";
import Calculator from "./Calculator";

const buttonsList = [
  "1",
  "2",
  "3",
  "4",
  "5",
  "6",
  "7",
  "8",
  "9",
  "0",
  "/",
  "x",
  "-",
  "+",
  "=",
  "AC",
];

test("render calculator", () => {
  const { container } = render(<Calculator />);

  const initialValue = screen.getByRole("heading", { name: "0" });
  expect(initialValue).toBeInTheDocument();

  expect(screen.getAllByRole("button")).toHaveLength(16);

  expect(container).toMatchSnapshot();
});

test.each(buttonsList)("renders %p button", (btn) => {
  render(<Calculator />);

  const getButton = screen.getByRole("button", { name: btn });

  expect(getButton).toBeInTheDocument();
});

test("when first clicked on action button", () => {
  render(<Calculator />);

  const result = screen.getByRole("heading", { name: "0" });

  const buttonPlus = screen.getByRole("button", { name: "+" });
  fireEvent.click(buttonPlus);

  expect(result).toBeInTheDocument();
});

test("functionality", () => {
  render(<Calculator />);

  const result = screen.getByRole("heading", { name: "0" });

  const buttonSix = screen.getByRole("button", { name: "6" });
  fireEvent.click(buttonSix);

  expect(result).toHaveTextContent("6");

  const buttonPlus = screen.getByRole("button", { name: "+" });
  fireEvent.click(buttonPlus);

  const buttonThree = screen.getByRole("button", { name: "3" });
  fireEvent.click(buttonThree);

  expect(result).toHaveTextContent("3");

  const buttonEquals = screen.getByRole("button", { name: "=" });
  fireEvent.click(buttonEquals);

  expect(result).toBeInTheDocument();
});

test("clear the calculator", async () => {
  render(<Calculator />);

  const buttonSix = screen.getByRole("button", { name: "6" });
  fireEvent.click(buttonSix);

  const buttonNine = screen.getByRole("button", { name: "9" });
  fireEvent.click(buttonNine);

  const result = screen.queryByText("69");
  expect(result).toBeInTheDocument();

  const buttonAC = screen.getByRole("button", { name: "AC" });
  fireEvent.click(buttonAC);

  await waitFor(() => {
    expect(screen.queryByText("69")).not.toBeInTheDocument();
  });

  expect(result).toHaveTextContent("0");
});
