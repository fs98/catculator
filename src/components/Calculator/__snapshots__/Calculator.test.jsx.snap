// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render calculator 1`] = `
<div>
  <div
    class="calculator container mx-auto md:w-1/2"
  >
    <h1
      class="p-8 text-right text-6xl"
    >
      0
    </h1>
    <div
      class="grid grid-cols-4 gap-2"
    >
      <button
        class="px-7 py-8 text-4xl text-center hover:cursor-pointer hover:trasform hover:-translate-y-2 active:translate-y-0 bg-gray-300 hover:bg-gray-400"
      >
        7
      </button>
      <button
        class="px-7 py-8 text-4xl text-center hover:cursor-pointer hover:trasform hover:-translate-y-2 active:translate-y-0 bg-gray-300 hover:bg-gray-400"
      >
        8
      </button>
      <button
        class="px-7 py-8 text-4xl text-center hover:cursor-pointer hover:trasform hover:-translate-y-2 active:translate-y-0 bg-gray-300 hover:bg-gray-400"
      >
        9
      </button>
      <button
        class="px-7 py-8 text-4xl text-center hover:cursor-pointer hover:trasform hover:-translate-y-2 active:translate-y-0 bg-orange-500 hover:bg-orange-600"
      >
        /
      </button>
      <button
        class="px-7 py-8 text-4xl text-center hover:cursor-pointer hover:trasform hover:-translate-y-2 active:translate-y-0 bg-gray-300 hover:bg-gray-400"
      >
        4
      </button>
      <button
        class="px-7 py-8 text-4xl text-center hover:cursor-pointer hover:trasform hover:-translate-y-2 active:translate-y-0 bg-gray-300 hover:bg-gray-400"
      >
        5
      </button>
      <button
        class="px-7 py-8 text-4xl text-center hover:cursor-pointer hover:trasform hover:-translate-y-2 active:translate-y-0 bg-gray-300 hover:bg-gray-400"
      >
        6
      </button>
      <button
        class="px-7 py-8 text-4xl text-center hover:cursor-pointer hover:trasform hover:-translate-y-2 active:translate-y-0 bg-orange-500 hover:bg-orange-600"
      >
        x
      </button>
      <button
        class="px-7 py-8 text-4xl text-center hover:cursor-pointer hover:trasform hover:-translate-y-2 active:translate-y-0 bg-gray-300 hover:bg-gray-400"
      >
        1
      </button>
      <button
        class="px-7 py-8 text-4xl text-center hover:cursor-pointer hover:trasform hover:-translate-y-2 active:translate-y-0 bg-gray-300 hover:bg-gray-400"
      >
        2
      </button>
      <button
        class="px-7 py-8 text-4xl text-center hover:cursor-pointer hover:trasform hover:-translate-y-2 active:translate-y-0 bg-gray-300 hover:bg-gray-400"
      >
        3
      </button>
      <button
        class="px-7 py-8 text-4xl text-center hover:cursor-pointer hover:trasform hover:-translate-y-2 active:translate-y-0 bg-orange-500 hover:bg-orange-600"
      >
        -
      </button>
      <button
        class="px-7 py-8 text-4xl text-center hover:cursor-pointer hover:trasform hover:-translate-y-2 active:translate-y-0 bg-orange-500 hover:bg-orange-600"
      >
        AC
      </button>
      <button
        class="px-7 py-8 text-4xl text-center hover:cursor-pointer hover:trasform hover:-translate-y-2 active:translate-y-0 bg-gray-300 hover:bg-gray-400"
      >
        0
      </button>
      <button
        class="px-7 py-8 text-4xl text-center hover:cursor-pointer hover:trasform hover:-translate-y-2 active:translate-y-0 bg-orange-500 hover:bg-orange-600"
      >
        =
      </button>
      <button
        class="px-7 py-8 text-4xl text-center hover:cursor-pointer hover:trasform hover:-translate-y-2 active:translate-y-0 bg-orange-500 hover:bg-orange-600"
      >
        +
      </button>
    </div>
  </div>
</div>
`;
